plugins/checksum/checksum.c

```diff
@@ -22,244 +22,431 @@
+
+  g_input_stream_read_async(G_INPUT_STREAM(file_stream), buffer, *buffer_size,
+                            G_PRIORITY_DEFAULT, NULL, stream_async_cb,
+                            (void *)user_data);
+}
+
+typedef struct {
+  async_result_cb *result_cb;
+  GFileInputStream *file_stream;
+  char *filename;
+  GFile *file;
+  gpointer bound_data;
+} file_async_cb_data;
+
+static void file_async_cb(gboolean result, gpointer user_data) {
+
+  file_async_cb_data *bound_vars = (void *)user_data;
+
+  if (!result) {
+    hexchat_printf(ph, "Checksum: Failed to generate checksum for %s\n",
+                   bound_vars->filename);
+  }
+
+  g_object_unref(bound_vars->file_stream);
+  g_object_unref(bound_vars->file);
+  async_result_cb *cb = bound_vars->result_cb;
+  gpointer bound_data = bound_vars->bound_data;
+  free(bound_vars);
+
+  return cb(result, bound_data);
 }
 
-static gboolean
-sha256_from_file (char *filename, char out_buf[])
-{
-	GFileInputStream *file_stream;
-	char *filename_fs;
-	GFile *file;
-
-	filename_fs = g_filename_from_utf8 (filename, -1, NULL, NULL, NULL);
-	if (!filename_fs)
-	{
-		hexchat_printf (ph, "Checksum: Invalid filename (%s)\n", filename);
-		return FALSE;
-	}
-
-	file = g_file_new_for_path (filename_fs);
-	g_free (filename_fs);
-	if (!file)
-	{
-		hexchat_printf (ph, "Checksum: Failed to open %s\n", filename);
-		return FALSE;
-	}
-
-	if (!check_limit (file))
-	{
-		hexchat_printf (ph, "Checksum: %s is larger than size limit. You can increase it with /CHECKSUM SET.\n", filename);
-		g_object_unref (file);
-		return FALSE;
-	}
-
-	file_stream = g_file_read (file, NULL, NULL);
-	if (!file_stream)
-	{
-		hexchat_printf (ph, "Checksum: Failed to read file %s\n", filename);
-		g_object_unref (file);
-		return FALSE;
-	}
-
-	if (!sha256_from_stream (file_stream, out_buf))
-	{
-		hexchat_printf (ph, "Checksum: Failed to generate checksum for %s\n", filename);
-		g_object_unref (file_stream);
-		g_object_unref (file);
-		return FALSE;
-	}
-
-	g_object_unref (file_stream);
-	g_object_unref (file);
-	return TRUE;
+static void sha256_from_file_async(char *filename, char out_buf[],
+                                   async_result_cb *result_cb,
+                                   gpointer bound_data) {
+  GFileInputStream *file_stream;
+  char *filename_fs;
+  GFile *file;
+
+  filename_fs = g_filename_from_utf8(filename, -1, NULL, NULL, NULL);
+  if (!filename_fs) {
+    hexchat_printf(ph, "Checksum: Invalid filename (%s)\n", filename);
+    return result_cb(FALSE, bound_data);
+  }
+
+  file = g_file_new_for_path(filename_fs);
+  g_free(filename_fs);
+  if (!file) {
+    hexchat_printf(ph, "Checksum: Failed to open %s\n", filename);
+    return result_cb(FALSE, bound_data);
+  }
+
+  if (!check_limit(file)) {
+    hexchat_printf(ph,
+                   "Checksum: %s is larger than size limit. You can increase "
+                   "it with /CHECKSUM SET.\n",
+                   filename);
```

Comment:
Since this is no longer blocking, I think we should remove this setting entirely.

It doesn't really matter too much how big it is since we do chunked reads.